uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php

```diff
@@ -1440,4 +1436,33 @@ private static function getCurrentAbsenceType($leaveRequestID) {
 
     return AbsenceType::findById($currentLeaveRequest->type_id);
   }
+
+  /**
+   * Get the number of working days for a leave request, Days with
+   * balance change of zero or that evaluates to false are either non
+   * working day, weekend or public holidays.
+   * The days for which the balance amounts are not zero are equal
+   * to the number of working days.
+   *
+   * @param array $params
+   *
+   * @return int
+   */
+  private static function getNumberOfWorkingDaysForLeaveRequest($params) {
+    $fromDate = new DateTime($params['from_date']);
+    $toDate = new DateTime($params['to_date']);
+    $balanceDetails = self::calculateBalanceChange(
```

Comment:
I wonder if we can do this without `calculateBalanceChange()`. This calculation is quite expensive and we are already calling it in the `validateEntitlementAndWorkingDayAndBalanceChange()`, this is just making the Leave Request validation slower than it already is. Besides, I'd say there are some readability and single responsibility issues here. The fact that we're using this `calculateBalanceChange()` method for two different types of validations, shows that it's doing different things. This is not good. In fact, we can see that when you do the whole balance calculation but in the end you only care about the number of balance changes where the amount is > 0. We need to improve this. Perhaps split that method into smaller ones

Reply:
Good Idea, Let me see what I can come up with. I wasn't too comfortable with the expensive call too, Will try split the method.

Reply:
I tried splitting the method and merging two validation methods that were using the `calculateBalanceChange()` for their validations. It seems to work fine, Only issue is that the validation method i merged it to i.e validateEntitlementAndWorkingDayAndBalanceChange does not get run when leave requests are imported and which subsequently means that validations for max consecutive days will not be ran either for that case.
What do you think?